plugins/hu.elte.txtuml.api.model/src/hu/elte/txtuml/api/model/ModelExecutorThread.java

```diff
@@ -76,24 +79,20 @@ private void newMailboxEntry(MailboxEntry entry) {
 	 * becomes empty.
 	 */
 	void shutdown() {
-		try {
-			class ShutdownAction implements MailboxEntry {
-				@Override
-				public void execute() {
-					if (mailbox.isEmpty()) {
-						ModelExecutor.shutdownNow();
-					} else {
-						try {
-							mailbox.put(new ShutdownAction());
-						} catch (InterruptedException e) {
-						}
-					}
-				}
-			}
+		shutdownWhenEmpty = true;

```

Comment:
But it may never notice that flag was set to true.

Furthermore, if it is waiting for the mailbox to have an entry, you wake it by giving it an entry. However, if it does not update the value of the flag and uses the cached value, then it enters the next iteration and waits again for the mailbox to get an entry. And will wait forever. Or at least, until you kill the JVM.